samples/matter/common/src/watchdog/watchdog.cpp

```diff
@@ -0,0 +1,224 @@
+	}
+
+	LOG_DBG("Registered a watchdog channel %d with feed interval %u ms", mChannel, mFeedingTime);
+}
+
+WatchdogSource::~WatchdogSource()
+{
+	if (k_timer_status_get(&mTimer)) {
+		k_timer_stop(&mTimer);
+	}
+
+	/* Stop the global timer */
+	AssertAndReturn(Disable(), "Watchdog cannot be disabled");
+
+	sys_snode_t *node;
+	sys_snode_t *tmpNodeSafe;
+	WatchdogListItem *item;
+	SYS_SLIST_FOR_EACH_NODE_SAFE (&sWatchdogList, node, tmpNodeSafe) {
+		item = reinterpret_cast<WatchdogListItem *>(node);
+		if (item && item->mWatchdog == this) {
+			/* Remove this object from the list */
+			sys_slist_find_and_remove(&sWatchdogList, item);
+			chip::Platform::Delete(item);
+		} else if (item) {
+			/* Restore all other Watchdog sources from the list */
+			item->mWatchdog->Restore();
+		}
+	}
+
+	/* Enable the global timer */
+	AssertAndReturn(Enable(), "Watchdog cannot be enabled");
+}
+
+void WatchdogSource::Feed()
+{
+	if (sEnabled) {
+		if (mChannel == kInvalidChannel) {
+			LOG_ERR("Call for Watchdog FEED, but the watchdog has not been initialized");
+		}
+
+		wdt_feed(wdt, mChannel);
+	}
+}
+
+bool WatchdogSource::Restore()
+{
+	if (sEnabled) {
+		return false;
+	}
+
+	/* instal wdt timeout based on the previous configuration */
+	wdt_timeout_cfg wdtConfig = {};
+	wdtConfig.flags = WDT_FLAG_RESET_SOC;
+	wdtConfig.window.min = 0;
+	wdtConfig.window.max = CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT;
+	wdtConfig.callback = nullptr;
+
+	mChannel = wdt_install_timeout(wdt, &wdtConfig);
+	if (-ENOMEM == mChannel) {
+		LOG_ERR("Cannot create a new Watchdog source because there is no more available Watchdog channels");
+		return false;
+	}
+
+	if (mChannel < 0) {
+		return false;
+	}
+
+	/* Start the feeding timer if the feeding callback exists */
+	if (mCallback) {
+		k_timer_start(&mTimer, K_MSEC(mFeedingTime), K_MSEC(mFeedingTime));
+	}
+
+	return true;
+}
+
+void WatchdogSource::TimerTimeoutCallback(k_timer *timer)
+{
+	if (sEnabled) {
+		WatchdogSource *watchdog = reinterpret_cast<WatchdogSource *>(k_timer_user_data_get(timer));
+		// Feeding using timer only of callback is set
+		if (watchdog && watchdog->mCallback) {
+			watchdog->mCallback(watchdog);
+		}
+	}
+}
+
+bool Nrf::Watchdog::Enable()
+{
+	/* Enable the global timer only of there is at least one source assigned */
+	if (sys_slist_is_empty(&sWatchdogList)) {
+		return false;
+	}
+
+	if (sEnabled || wdt_setup(wdt, WatchdogSource::kWatchdogOptions) < 0) {
+		return false;
+	}
+
+	LOG_DBG("Global watchdog started");
+	sEnabled = true;
+	return true;
+}
+
+bool Nrf::Watchdog::Disable()
+{
+	if (!sEnabled) {
+		return false;
+	}
+
+	int ret = wdt_disable(wdt);
+	if (ret != 0) {
+		return false;
+	}
+
+	/* Stop all feeding timers*/
+	sys_snode_t *node;
+	sys_snode_t *tmpNodeSafe;
+	WatchdogListItem *item;
+	SYS_SLIST_FOR_EACH_NODE_SAFE (&sWatchdogList, node, tmpNodeSafe) {
+		item = reinterpret_cast<WatchdogListItem *>(node);
+		if (k_timer_status_get(&item->mWatchdog->mTimer)) {
```

Comment:
Why cancelling the timer only if it expired at least once?

Reply:
That's wrong, I need to stop the timer only if after the node removal there is no one on the list anymore.